stdlib/public/runtime/ImageInspectionCommon.h

```diff
@@ -57,14 +57,27 @@ struct SectionInfo {
   const char *data;
 };
 
-// Called by injected constructors when a dynamic library is loaded.
+/// Called by injected constructors when a dynamic library is loaded.
+///
+/// \param sections A structure describing the metadata sections in the
+///     newly-loaded image.
+///
+/// \warning The runtime keeps a reference to \a sections and may mutate it, so
+///   it \em must be mutable and long-lived (that is, statically or dynamically
+///   allocated.) The effect of passing a pointer to a local value is undefined.
 SWIFT_RUNTIME_EXPORT
-void swift_addNewDSOImage(const void *addr);
+void swift_addNewDSOImage(struct swift::MetadataSections *sections);
 
 #ifndef NDEBUG
 
 SWIFT_RUNTIME_EXPORT
-const char *swift_getMetadataSectionName(void *metadata_section);
+const char *
+swift_getMetadataSectionName(const struct swift::MetadataSections *section);
+
+SWIFT_RUNTIME_EXPORT
+void swift_getMetadataSectionBaseAddress(
+  const struct swift::MetadataSections *section,
+  void const **out_actual, void const **out_expected);
```

Comment:
Please guard this with `__cplusplus` or take the section metadata as a void pointer.

Reply:
This header is already not C-compatible, so this should not be an issue.